### PR TITLE
Implement support for v4 presets files

### PR DESCRIPTION
**Pull request by chausner**
## This change addresses items #2492 and #2469

The primary addition is support for version 4 presets files.

The following changes are proposed:

- A JSON schema for version 4 is added.

  I noticed that the official schema that Kitware publishes is in a somewhat different format than what is included in this extension. The official one makes more use of `$ref`s while the one in the extension only uses `$ref`s sparingly and also includes additional schema information for vendor-specific attributes.

  In order to update the schema to v4, I started with the v3 one in the extension and diffed it with the official v4 one.

- Support for the `${fileDir}` macro which is new in v4 is added.
- Logic is added to support `include`s in preset files.

  The implementation approach I've chosen is to recursively parse any included presets files and merge them into a single instance of `preset.PresetsFile`. It seemed the easiest way that requires little changes in other places but I would be happy to hear if there is a better way.

  Correct implementation of `${fileDir}` and `include` requires that `preset.Preset` tracks the file where the preset was originally defined (`$fileDir` should resolve to the directory of the include where that macro is used, not the root preset file. It is different, however, if a preset inherits from a preset in an included file: in this case, `$fileDir` should resolve to the directory of the presets file that inherits). For this purpose, I added some helper fields `preset.PresetsFile.__path` and `preset.Preset.__file`.

  The file watcher logic is extended to also watch for any changes to included presets files.

- A fix for #2469. The fix makes use of the helper field `preset.Preset.__file` which allows to query the version of the presets file where a preset is defined.

### CODE REVIEW

File: src/presetsController.ts [hunk 1]
```diff
@@ -955,17 +958,88 @@ export class PresetsController {
...
+
+            // Do not include files more than once
+            if (mergedFiles.has(fullIncludePath)) {
+                continue;
+            }
+            mergedFiles.add(fullIncludePath);
+
+            const includeFileBuffer = await this.readPresetsFile(fullIncludePath);
+            if (!includeFileBuffer) {
+                log.error(localize('included.presets.file.not.found', 'Included presets file {0} cannot be found', fullIncludePath));
```
Comment (chausner) [inline src/presetsController.ts:996]: I introduced this new localized string. This would need to be localized and added to the translation files.

**Comment by chausner** [comment]
@bobbrow I would like to do more testing before declaring this PR ready to be reviewed but maybe you or somebody else would like to have an early look and let me know if the approach looks fine to you so far.

**Comment by chausner** [comment]
#2492 mentions `resolvePackageReferences` which is new in v4. I don't think anything needs to be added in the extension for this (apart from the schema addition). Can somebody confirm?

**Comment by Zingam** [comment]
There is a V5 now.

**Comment by bobbrow** [comment]
> There is a V5 now.

Yes, we know that's coming in 3.24.  We're trying to release 1.11 and had some delays with the cpptools v6 integration PR.  We'll work on getting this v4 PR in shortly after for 1.12 pre-release.  Then we can do v5 after that.  v5 doesn't look like it has too much in it so far (a test truncation property and path separator variable)

**Review by bobbrow — COMMENTED** [review]
I'm sharing a few comments to take a look at while we review this.

We'll add some comments to the schema file as well.  There were a few changes we might want to make there.

And it would be great if you could add some tests for the new `includes` feature somewhere.

Thank you for your work on this!

File: src/presetsController.ts [hunk 2]
```diff
@@ -1028,12 +1080,23 @@ export class PresetsController {
...
-    dispose() {
+    private async watchPresetsChange() {
         if (this._presetsWatcher) {
             this._presetsWatcher.close().then(() => {}, () => {});
         }
-        if (this._userPresetsWatcher) {
-            this._userPresetsWatcher.close().then(() => {}, () => {});
+
+        const presetsChanged = async () => {
+            await this.reapplyPresets();
```
[the remark this answers is not in the document]
Comment (chausner) [reply src/presetsController.ts]: Done in ba726b3d0eceb77cd314701caf4f1aae8c269641.

File: src/presetsController.ts [hunk 3]
```diff
@@ -955,17 +935,89 @@ export class PresetsController {
         return presetsFile;
     }
 
+    private populatePrivatePresetsFields(presetsFile: preset.PresetsFile | undefined, file: string) {
+        if (!presetsFile) {
+            return;
+        }
+
+        presetsFile.__path = file;
+        const setFile = (presets: preset.Preset[] | undefined) => {
```
[the remark this answers is not in the document]
Comment (chausner) [reply src/presetsController.ts]: Done in ba726b3d0eceb77cd314701caf4f1aae8c269641.

File: src/preset.ts [hunk 4]
```diff
@@ -18,9 +18,12 @@ const log = logging.createLogger('preset');
 export interface PresetsFile {
     version: number;
     cmakeMinimumRequired?: util.Version;
+    include?: string[] | undefined;
```
[the remark this answers is not in the document]
Comment (chausner) [reply src/preset.ts]: Done in ba726b3d0eceb77cd314701caf4f1aae8c269641.

File: src/presetsController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (chausner) [reply src/presetsController.ts]: On another note, I wonder if there should be semaphore limiting the number of concurrent calls to `reapplyPresets` to 1. I imagine the file system watcher might invoke the callbacks more than once while `reapplyPresets` hasn't finished yet.

File: src/presetsController.ts [hunk 5]
```diff
@@ -119,7 +89,7 @@ export class PresetsController {
             presetsController._sourceDir = await expandSourceDir(value);
 
             if (presetsController._sourceDir !== oldSourceDir) {
-                await watchPresetsChange();
+                await presetsController.watchPresetsChange();
```
[the remark this answers is not in the document]
Comment (chausner) [reply src/presetsController.ts:93]: Good catch! I unintentionally removed that call when I refactored that part. Fixed in 5739462cc695568a3021388d90f509a0cd5c0874.

**Comment by chausner** [comment]
> And it would be great if you could add some tests for the new includes feature somewhere.

I had a brief look into the tests folder but only found tests for some parsing functions but none for the overall preset functionality: [test/unit-tests/presets.test.ts](https://github.com/microsoft/vscode-cmake-tools/blob/main/test/unit-tests/presets.test.ts).

File: src/presetsController.ts (hunk 1 above)
Comment (bobbrow) [reply src/presetsController.ts:975]: This will happen automatically after the PR is committed.

**Comment by bobbrow** [comment]
> I had a brief look into the tests folder but only found tests for some parsing functions but none for the overall preset functionality: [test/unit-tests/presets.test.ts](https://github.com/microsoft/vscode-cmake-tools/blob/main/test/unit-tests/presets.test.ts).

I believe we have more coverage in the `test/extension-tests/single-root-UI` folder for presets, but I forget how good it is.  If it's easy enough to add some basic coverage for the includes, we would appreciate it. Otherwise, we can investigate adding some more afterward.

**Comment by chausner** [comment]
> I believe we have more coverage in the `test/extension-tests/single-root-UI` folder for presets, but I forget how good it is. If it's easy enough to add some basic coverage for the includes, we would appreciate it. Otherwise, we can investigate adding some more afterward.

Thanks. I added a basic test for now in 8586780cb65d9da56bfa550543b7f2065e6221d6. I haven't been able to test it locally, though, because running `npm run extensionTestsSingleRoot` shows all tests failing and there are warnings:
> [vscode-cpptools-api] C/C++ extension is not installed

Do I need to run anything else before? I do have the C/C++ extension installed.

**Comment by bobbrow** [comment]
I tried running the test multiple times and am unable to reproduce the error on linux locally.  If you rename `include-presets.test.ts` to `include-presets.ts` that should disable the test and we can open a new issue to ensure we add tests for it (though I think the existing tests will actually get some coverage of it since the CMakeUserPresets.json was changed.

**Comment by bobbrow** [comment]
I'm trying one last idea to exclude that generated file from the test build.  If that doesn't work, let's disable the test for now.

**Comment by bobbrow** [comment]
Thank you so much for your help with this!

**Comment by prince-chrismc** [comment]
I was having issues and eventually I discovered this PR... I download Version: 1.12.0-1ca9c49 (from main's action results) and the v4 presets information I need instantly started working 💟 

Thanks for all you hard work!

**Comment by bobbrow** [comment]
@prince-chrismc Every morning we publish a build to the VS Code Marketplace if anything was submitted to main the day before.  From the extensions panel, you can right click on the CMake Tools item and select "Switch to Pre-Release Version". This feature is available starting in version 1.12.3.

**Comment by prince-chrismc** [comment]
Thanks for the tip!

**Comment by mpusz** [comment]
I just tested the new feature. It mostly works great!

The only problem I noticed is that the Build Presets are not refreshed when the included file is changed/appended with new configurations.

For example:
1. I generated a Release configuration with Conan
2. Loaded CMakeUserPresets.json in VSCode
3. Realized that the Debug is missing so I generated it with Conan which added a proper configuration to the previously included CMakePresets.json file
4. I needed to restart the CMake Tools extension state to see the update of presets.

**Comment by chausner** [comment]
In theory, it should detect the file change and I think I tested the scenario with Conan but let me check again and see if I can reproduce it.
________________________________

**Comment by chausner** [comment]
@mpusz I was able to reproduce the issue and found the problem. I submitted a fix here: https://github.com/microsoft/vscode-cmake-tools/pull/2619.

Thanks again for reporting the problem.

**Pull request merged**